### PR TITLE
CSE-337: Support for Builder - impTest Test Framework Improvements

### PR DESCRIPTION
**Pull request by dbukhmastov**

### CODE REVIEW

File: src/lib/Commands/InitCommand.js [hunk 1]
```diff
@@ -119,10 +143,10 @@ class InitCommand extends AbstractCommand {
     for (const modelId in this._models) {
       for (const deviceId of  this._models[modelId].devices) {
         table.push([
-          deviceId,
-          this._devices[deviceId].name,
-          this._models[modelId].id,
-          this._models[modelId].name,
+          deviceId || "",
```
Comment (snork-nbl) [inline src/lib/Commands/InitCommand.js]: why is that?

File: src/lib/Commands/TestCommand/index.js [hunk 2]
```diff
@@ -281,9 +310,28 @@ class TestCommand extends AbstractCommand {
...
+        var propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop + '\\s*\\}', 'g'); // default regexp
+        if (prop.match(UNSUPPORTED_SYMBOLS_REGEXP)) { // replace unsupported symbols in name
+          propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, '.') + '\\s*\\}', 'g');
+          prop = prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X');
+        }
+        if (testCode.match(propertyRegExp)) {
+           testCode = testCode.replace(propertyRegExp, '@{' + prop + '}');
+           if (propValue.match(UNSUPPORTED_SYMBOLS_REGEXP)) {  // replace unsupported symbols in value
+             let newPropValue = propValue.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X');
+             usupportedValues[newPropValue] = propValue; // store originalvalue
```
Comment (snork-nbl) [inline src/lib/Commands/TestCommand/index.js]: usupportedValues-> unsupportedValues

File: src/lib/Commands/TestCommand/index.js [hunk 3]
```diff
@@ -281,9 +310,28 @@ class TestCommand extends AbstractCommand {
...
+        var propValue = process.env[prop].replace(new RegExp('\\\\', 'g'), "\\\\");
+        var propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop + '\\s*\\}', 'g'); // default regexp
+        if (prop.match(UNSUPPORTED_SYMBOLS_REGEXP)) { // replace unsupported symbols in name
+          propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, '.') + '\\s*\\}', 'g');
+          prop = prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X');
+        }
+        if (testCode.match(propertyRegExp)) {
+           testCode = testCode.replace(propertyRegExp, '@{' + prop + '}');
+           if (propValue.match(UNSUPPORTED_SYMBOLS_REGEXP)) {  // replace unsupported symbols in value
+             let newPropValue = propValue.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X');
```
Comment (snork-nbl) [inline src/lib/Commands/TestCommand/index.js]: Why X?

File: src/lib/Commands/InitCommand.js (hunk 1 above)
Comment (dbukhmastov) [reply src/lib/Commands/InitCommand.js]: The change has been made In the electricimp/impTest develop branch: https://github.com/electricimp/impTest/commit/b64e9c4b64f0b533015c382ddc70653e23aa065d

Comment (ppetrosh) [reply src/lib/Commands/InitCommand.js]: Yep I remember it looks familiar. So I was wondering where it came from ) Thanks for reminding.

Please leave these changes, just alight them so that it looks better please:

```javascript
        table.push([
          deviceId                     || "",
          this._devices[deviceId].name || "",
          this._models[modelId].id     || "",
          this._models[modelId].name   || "",
          this._devices[deviceId].powerstate === 'online'
            ? c.green(this._devices[deviceId].powerstate)
            : c.red(this._devices[deviceId].powerstate)
        ]);
```

File: src/lib/Commands/TestCommand/index.js [hunk 4]
```diff
@@ -49,6 +72,16 @@ const DEFAULT_EXTRA_TEST_MESSAGE_TIMEOUT = 5;
 const BUILD_API_KEY_ENV_VAR = 'IMP_BUILD_API_KEY';
 
 /**
+ * Unsupported in Builder symbols
+ */
+const UNSUPPORTED_SYMBOLS_REGEXP = /[^_A-Za-z0-9]/g;
+
+/**
+ * Unsupported in Builder symbols
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Comment doesn't match the code here.

File: src/lib/Commands/TestCommand/index.js [hunk 5]
```diff
@@ -281,9 +310,28 @@ class TestCommand extends AbstractCommand {
...
     // read/process test code
-    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim();
-    this._codeProcessor.variables.__FILE__ = path.basename(testFile.path);
-    testCode = this._codeProcessor.process(testCode);
+    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim().replace(LINE_AND_FILE_REGEXP, "@{__");
+    let unsupportedValues = {}; // map of syntetic_value : original_value
+    var environmentVars = "";
+    for (var prop in process.env) {
+      if (prop !== BUILD_API_KEY_ENV_VAR) { //deny to access for BUILD_API_KEY_ENV_VAR
+        var propValue = process.env[prop].replace(new RegExp('\\\\', 'g'), "\\\\");
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Two questions here:

1. Why do you need to escape the characters here at all?
1. `new RegExp('\\\\', 'g')` -> `/\\\\/g` ?
1. why not `replace(/\\/g, "\\\\")` ?
1. what about other special chars:
```
str.replace(/\\/g, "\\\\")
    .replace(/\$/g, "\\$")
    .replace(/'/g, "\\'")
    .replace(/"/g, "\\\"");
```

File: src/lib/Commands/TestCommand/index.js [hunk 6]
```diff
@@ -49,6 +72,16 @@ const DEFAULT_EXTRA_TEST_MESSAGE_TIMEOUT = 5;
 const BUILD_API_KEY_ENV_VAR = 'IMP_BUILD_API_KEY';
 
 /**
+ * Unsupported in Builder symbols
+ */
+const UNSUPPORTED_SYMBOLS_REGEXP = /[^_A-Za-z0-9]/g;
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Why are not these characters supported?

File: src/lib/Commands/TestCommand/index.js [hunk 7]
```diff
@@ -281,9 +310,28 @@ class TestCommand extends AbstractCommand {
...
+      if (prop !== BUILD_API_KEY_ENV_VAR) { //deny to access for BUILD_API_KEY_ENV_VAR
+        var propValue = process.env[prop].replace(new RegExp('\\\\', 'g'), "\\\\");
+        var propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop + '\\s*\\}', 'g'); // default regexp
+        if (prop.match(UNSUPPORTED_SYMBOLS_REGEXP)) { // replace unsupported symbols in name
+          propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, '.') + '\\s*\\}', 'g');
+          prop = prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X');
+        }
+        if (testCode.match(propertyRegExp)) {
+           testCode = testCode.replace(propertyRegExp, '@{' + prop + '}');
+           if (propValue.match(UNSUPPORTED_SYMBOLS_REGEXP)) {  // replace unsupported symbols in value
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Why do you need to do the replacement for values?

File: src/lib/Commands/TestCommand/index.js [hunk 8]
```diff
@@ -315,17 +363,20 @@ imp.wakeup(${this.startupDelay /* prevent log sessions mixing, allow service mes
     // quote file name for line control statement
     const quoteFilename = f => f.replace('"', '\\"');
 
+    let tmpFrameworkFile = this.testFrameworkFile.replace(new RegExp('\\\\', 'g'),"/");
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: 1. Will this work on Windows?
2. `new RegExp('\\\\', 'g')`->`/\\\\/g`

File: src/lib/Commands/TestCommand/index.js [hunk 9]
```diff
@@ -337,10 +388,18 @@ ${bootstrapCode}
 }
 
 // resolve modules
-__module_tests(__module_ImpUnit_Promise_exports, __module_impUnit_exports.ImpTestCase);
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js:340]: I believe we still need to pre-include Promise here as many tests rely on it.

File: src/lib/Commands/TestCommand/index.js [hunk 10]
```diff
@@ -717,8 +785,6 @@ ${(this._sourceCode.agent || '/* no agent source */')}
         }
 
         this._agentSource = fs.readFileSync(sourceFilePath, 'utf-8').trim();
-        this._codeProcessor.variables.__FILE__ = path.basename(sourceFilePath);
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Why is that removed.

File: src/lib/Commands/TestCommand/index.js (hunk 6 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: The Builder support letters, numders and "_" the  in name: @set NAME VALUE

File: src/lib/Commands/TestCommand/index.js (hunk 7 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: The approach is to replace old style #{env:<SOME_NAME>} with @{CORRECTED_SOME_NAME} and to add approprite @set CORRECTED_SOME_NAME "CORRECTED_VALUE"
The Builder doesn't support `\` AND `\"` in <value:expression> on win7 Additionally absent any doc about symbol applicability in <value:expression>. So I preffer (I can't predict which symbols isn't supported by Builder) to repleace ALL "strange" symbols. 


One more approach to find unused in source code text like
var unusedText = ('A'+Math.random()).replace('.', '');
while (testCode.match(new RegExp(unusedText+"[0-9]*"))) {
  unusedText = ('A'+Math.random()).replace('.', '');
}

var index = 0; (unusedText+index)  <--- use this as name and as corrected value (after index++) like:
replace #{env:<SOME_NAME>} with @{(unusedText + index)} 
add @set (unusedText + index) "(unusedText + index++)"
after applaing Builder replace (unusedText + index++) with original value

File: src/lib/Commands/TestCommand/index.js (hunk 9 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js:340]: The Promise is in impUnit\index.nut  ->  #require "promise.class.nut:3.0.0" so the Promise 3.0.0 is present accessible in code as Promise and as ::Promise. Promise, JSONEncoder and Predix tests have been executed.

Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js:340]: The one issue can be -> the isPending() method was in old (removed) Promise AND It is absent in promise.class.nut:3.0.0

File: src/lib/Commands/TestCommand/index.js (hunk 4 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: Done

File: src/lib/Commands/TestCommand/index.js (hunk 8 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: 1) The code is tested on win7, all work fine
2) I can't say why, but the `/\\\\/g` doesn't work on win7:
Including source "C:sersmitryppDataoaming
pm
ode_modulesmptestrcli/../impUnit/index.nut"

DONE: it has been repleaced with `/\\/g`

File: src/lib/Commands/TestCommand/index.js (hunk 5 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: At the moment  I think that we have not to correct the original value. We can't predict all usecases.
So I would like to delete any env value corrections. What do you think? (additionallly see about unused in source code text belove)

In current version:
CommonProgramFiles=C:\Program Files\Common Files
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
The code is 
        `local myVar = "#{env:CommonProgramFiles}";`
        `myVar = "#{env:CommonProgramFiles(x86)}";`
The result of impTest working is:
        `local myVar = "C:\\Program Files\\Common Files";`
        `myVar = "C:\\Program Files (x86)\\Common Files";`
2, 3 and 4 seems it is better do not correct original value

File: src/lib/Commands/TestCommand/index.js (hunk 10 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: _codeProcessor is the "preprocessor". It has been replaced with Builder

File: src/lib/Commands/TestCommand/index.js (hunk 5 above)
Comment (ppetrosh) [reply src/lib/Commands/TestCommand/index.js]: To be honest I still don't see why we need to replace anything at all.

File: src/lib/Commands/TestCommand/index.js (hunk 6 above)
Comment (ppetrosh) [reply src/lib/Commands/TestCommand/index.js]: Are we aware of any other characters used in the existing tests? I would really doubt it. Even so, it's better to fix the tests rather than to invent a generic solution for this case here.

File: src/lib/Commands/TestCommand/index.js (hunk 7 above)
Comment (ppetrosh) [reply src/lib/Commands/TestCommand/index.js]: As above I don't believe we need this.

File: src/lib/Commands/TestCommand/index.js (hunk 9 above)
Comment (ppetrosh) [reply src/lib/Commands/TestCommand/index.js:340]: If `isPending` is ever used in the text sources, then the tests will need to be updated. So it shouldn't be an issue.

File: src/lib/Commands/TestCommand/index.js (hunk 10 above)
Comment (ppetrosh) [reply src/lib/Commands/TestCommand/index.js]: Got it.

File: src/lib/Commands/TestCommand/index.js (hunk 5 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: OK, it has been removed. The env prop values will be shipped as is.

File: src/lib/Commands/TestCommand/index.js (hunk 7 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: agree, it has been removed. The env prop values will be shipped as is.

File: src/lib/Commands/TestCommand/index.js (hunk 3 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: The env prop values will be shipped as is.

File: src/lib/Commands/InitCommand.js (hunk 1 above)
Comment (dbukhmastov) [reply src/lib/Commands/InitCommand.js]: Done

File: src/lib/Commands/TestCommand/index.js [hunk 11]
```diff
@@ -10,54 +33,56 @@ const fs = require('fs');
...
+// Name for BuildAPI key env var
+
 const BUILD_API_KEY_ENV_VAR = 'IMP_BUILD_API_KEY';
 
-/**
- * Test command
- */
+// Unsupported in Builder symbols
+
+const UNSUPPORTED_SYMBOLS_REGEXP = /[^_A-Za-z0-9]/g;
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: I still believe you don't need to replace any symbols in the variable names/values. Are there any use cases when it's needed?

File: src/lib/Commands/TestCommand/index.js [hunk 12]
```diff
@@ -268,22 +285,32 @@ class TestCommand extends AbstractCommand {
...
-    this._codeProcessor.variables.__FILE__ = path.basename(testFile.path);
-    testCode = this._codeProcessor.process(testCode);
+    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim().replace(LINE_AND_FILE_REGEXP, "@{__");
+    var environmentVars = "";
+    for (var prop in process.env) {
+      if (prop !== BUILD_API_KEY_ENV_VAR) { //deny to access for BUILD_API_KEY_ENV_VAR
+        let propValue = process.env[prop];
+        prop = prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X');
+        // Replace #{env:...} with @{...} if it is needed
+        let propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, '.') + '\\s*\\}', 'g');
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: You already replaced all these symbols to Xes on the line above. Here it makes no sense any more.

File: src/lib/Commands/TestCommand/index.js (hunk 11 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: WIN7 has some predefined environment props like CommonProgramFiles(x86) with unsupported symbols
if the #{env:CommonProgramFiles(x86)} is used we will generate follow code @set CommonProgramFiles(x86) "....." The Builder will fail.

Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: But I think that the approach should be the same as with properties values. So prop = prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X'); has been removed

File: src/lib/Commands/TestCommand/index.js (hunk 12 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: prop = prop.replace(UNSUPPORTED_SYMBOLS_REGEXP, 'X'); has been removed

File: src/lib/Commands/TestCommand/index.js (hunk 11 above)
Comment (ppetrosh) [reply src/lib/Commands/TestCommand/index.js]: I can't imagine any tests using the Windows specific variables.

Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: Removed

File: src/lib/Commands/TestCommand/index.js [hunk 13]
```diff
@@ -268,22 +281,30 @@ class TestCommand extends AbstractCommand {
...
 
-    /* [info] */
+    // [info]
     this._info(c.blue('Using ') + testFile.type + c.blue(' test file ') + testFile.name);
 
     // read/process test code
-    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim();
-    this._codeProcessor.variables.__FILE__ = path.basename(testFile.path);
-    testCode = this._codeProcessor.process(testCode);
+    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim().replace(LINE_AND_FILE_REGEXP, "@{__");
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Cool the code looks much better and cleaner to me know! Thanks.

The only think I would do is to explicitly do the replacement for `__FILE__`/`__PATH__`, rather than for anything that starts with __.

File: src/lib/Commands/TestCommand/index.js [hunk 14]
```diff
@@ -268,22 +281,30 @@ class TestCommand extends AbstractCommand {
...
+    var environmentVars = "";
+    for (var prop in process.env) {
+      if (prop !== BUILD_API_KEY_ENV_VAR) { //deny to access for BUILD_API_KEY_ENV_VAR
+        // Replace #{env:...} with @{...} if it is needed
+        let propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop + '\\s*\\}', 'g');
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: What if there is no such variable specified in the environment, but it's used in the code? How will user know about this issue?

File: src/lib/Commands/TestCommand/index.js (hunk 13 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: DONE - use separate regexp for line and file

File: src/lib/Commands/TestCommand/index.js (hunk 14 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: DONE: warning is added. Format is:
[+31.02/0.00s warning] Can't replace: #{env:PREDIX_UAA_URL},#{env:PREDIX_CLIENT_ID},#{env:PREDIX_CLIENT_SECRET},#{env:PREDIX_ASSET_URL},#{env:PREDIX_ASSET_ZONE_ID},#{env:PREDIX_TIME_SERIES_INGEST_URL},#{env:PREDIX_TIME_SERIES_ZONE_ID}

**Comment by ppetrosh** [comment]
In production Builder debug messages:

```
Including source "/working/cse/impTest-nbl/src/cli/../impUnit/index.nut"
Reading local file "/working/cse/impTest-nbl/src/impUnit/index.nut"
Including source "/working/cse/impTest-nbl/src/impUnit/lib/JSONEncoder.mod.nut"
Reading local file "/working/cse/impTest-nbl/src/impUnit/lib/JSONEncoder.mod.nut"
Including source "/working/cse/impTest-nbl/src/impUnit/src/ImpUnit.nut"
Reading local file "/working/cse/impTest-nbl/src/impUnit/src/ImpUnit.nut"
Including source "/working/cse/impTest-nbl/src/impUnit/src/Message.nut"
Reading local file "/working/cse/impTest-nbl/src/impUnit/src/Message.nut"
Including source "/working/cse/impTest-nbl/src/impUnit/src/TestCase.nut"
Reading local file "/working/cse/impTest-nbl/src/impUnit/src/TestCase.nut"
Including source "/working/cse/impTest-nbl/src/impUnit/src/TestRunner.nut"
Reading local file "/working/cse/impTest-nbl/src/impUnit/src/TestRunner.nut"
```

should not show up in the test run console.

this is what fixes it:

```
@@ -806,6 +806,14 @@ ${agentIncludeOrComment}
   get _Builder() {
     if (!this.__Builder) {
       this.__Builder = new Builder();
+      this.__Builder.logger = {
+        debug: function () {
+        },
+        info: function () {
+        },
+        warning: console.warning,
+        error: console.error
+      };
     }
     return this.__Builder;
   }
```

**Comment by ppetrosh** [comment]
Can you please also add a debug flag to dump preprocessed code into `build/preprocessed.[device|agent].nut`?

**Comment by ppetrosh** [comment]
I tried the develop branch and it looks like Builder isn't used for the test code preprocessing: a simple command `@set MYVAR 3` makes the compiler failing:

`[+0.94/0.48s error] Build API error "CompileFailed": Agent code compile failed`

File: src/lib/Commands/TestCommand/index.js [hunk 15]
```diff
@@ -268,22 +282,35 @@ class TestCommand extends AbstractCommand {
...
+    // [info]
     this._info(c.blue('Using ') + testFile.type + c.blue(' test file ') + testFile.name);
 
     // read/process test code
-    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim();
-    this._codeProcessor.variables.__FILE__ = path.basename(testFile.path);
-    testCode = this._codeProcessor.process(testCode);
+    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim()
+        .replace(LINE_REGEXP, "@{__LINE__}").replace(FILE_REGEXP, "@{__FILE__}");
+    var environmentVars = "";
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: `let` is better here

File: src/lib/Commands/TestCommand/index.js [hunk 16]
```diff
@@ -268,22 +282,35 @@ class TestCommand extends AbstractCommand {
...
+    for (var prop in process.env) {
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: again, `let`

File: src/lib/Commands/TestCommand/index.js [hunk 17]
```diff
@@ -268,22 +282,35 @@ class TestCommand extends AbstractCommand {
...
+      if (prop !== BUILD_API_KEY_ENV_VAR) { //deny to access for BUILD_API_KEY_ENV_VAR
+        // Replace #{env:...} with @{...} if it is needed
+        let propertyRegExp = new RegExp('\\#\\{env:\\s*' + prop + '\\s*\\}', 'g');
+        if (testCode.match(propertyRegExp)) {
+          testCode = testCode.replace(propertyRegExp, '@{' + prop + '}');
+          environmentVars = environmentVars + "@set "+ prop + " \"" + process.env[prop] + "\"\n";
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: and to be honest I believe we need to make Builder retrieving variables (if not set explicitly set in the code) from the environment, and avoid this dancing around the @set prop here. and it should be builder that swear on that

this needs to be discussed and thought over a little bit more and is probably worth filing an issue against the Builder.

BTW why don't you just replace the #{env:VARIABLE} with just it's "VALUE" instead of jumping up with @set/@{} expressions?

File: src/lib/Commands/TestCommand/index.js [hunk 18]
```diff
@@ -370,19 +403,42 @@ ${bootstrapCode}
...
+      __PATH__: path.dirname(testFile.path)
+    });
+
+    if (this.debug) {
+      // create folder to dump preprocessed code
+      if (!fs.existsSync('./build')) {
+        fs.mkdirSync('./build');
+      }
+      // write dump preprocessed codes
+      fs.writeFile('./build/preprocessed.agent.nut', agentCode, (err) => {
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js]: Please use the `preprocessed.<test file>` name to keep sources for all the tests, not just the last one executed.

**Comment by ppetrosh** [comment]
If you fix my recent comments/questions I believe I'll be good to approve it. It looks much better to me now. Thanks!

File: src/lib/Commands/TestCommand/index.js (hunk 18 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: DONE. But relative path is used

File: src/lib/Commands/TestCommand/index.js (hunk 15 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: DONE. environmentVars has been removed

File: src/lib/Commands/TestCommand/index.js (hunk 16 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js]: DONE

File: src/lib/Commands/TestCommand/index.js (hunk 17 above)
Comment (dbukhmastov) [inline src/lib/Commands/TestCommand/index.js]: "why don't you just replace" - Yes this is possible.

File: .gitmodules [hunk 19]
```diff
@@ -1,3 +1,3 @@
 [submodule "src/impUnit"]
 	path = src/impUnit
-	url = https://github.com/electricimp/impUnit.git
+	url = https://github.com/electricimp/impUnit.git
```
Comment (ppetrosh) [inline .gitmodules]: end of line's missing

File: src/lib/Commands/TestCommand/index.js [hunk 20]
```diff
@@ -268,22 +282,34 @@ class TestCommand extends AbstractCommand {
...
 
-    /* [info] */
+    // [info]
     this._info(c.blue('Using ') + testFile.type + c.blue(' test file ') + testFile.name);
 
     // read/process test code
-    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim();
-    this._codeProcessor.variables.__FILE__ = path.basename(testFile.path);
-    testCode = this._codeProcessor.process(testCode);
+    let testCode = fs.readFileSync(testFile.path, 'utf-8').trim()
```
Comment (ppetrosh) [inline src/lib/Commands/TestCommand/index.js:297]: As discussed offline, can you please add TODO here:

`TODO: Don't iterate over all the possible env variables, check only those that are used in the tests. If anything's used but is missing in the environment, raise a warning.`

**Review by ppetrosh — APPROVED** [review]

File: .gitmodules (hunk 19 above)
Comment (dbukhmastov) [reply .gitmodules]: Original content has been returned.

File: src/lib/Commands/TestCommand/index.js (hunk 20 above)
Comment (dbukhmastov) [reply src/lib/Commands/TestCommand/index.js:297]: todo has been added
